### PR TITLE
Fix #92 - Login Issue

### LINKED ISSUE #92: Login: no CSRF Token found
Hello,

What's this problem?

safaribooks.py --cred "saadaati@gmail.com:xxxxx" 9781492057284 --no-kindle

[-] Logging into Safari Books Online...
                                                                                                                        [#] Login: no CSRF Token found in the page. Unable to continue the login. Try again...
                                                                                                                        [+] Please delete all the `<BOOK NAME>/OEBPS/*.xhtml` files and restart the program.
                                                                                                                        [!] Aborting...

### PR DESCRIPTION
**Pull request by joshmaerk**
Change Login procedure from cookies based to session based.

### CODE REVIEW

File: safaribooks.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def qsize(self):
 
 class SafariBooks:
 
-    LOGIN_URL = SAFARI_BASE_URL + "/accounts/login/"
```
Comment (joshmaerk) [inline safaribooks.py]: Change Login URL to new Auth Path

File: safaribooks.py [hunk 2]
```diff
@@ -275,6 +275,8 @@ def __init__(self, args):
 
         self.cookies = {}
 
+        self.session = None
+
```
Comment (joshmaerk) [inline safaribooks.py]: Create a session variable to store session after login and use it for all following requests

File: safaribooks.py [hunk 3]
```diff
@@ -384,12 +386,15 @@ def update_cookies(self, jar):
...
-                headers=self.return_headers(url),
-                data=data,
-                **kwargs
-            )
+            if not self.session:
+                self.get_session(data=data)
+
+            response = getattr(self.session, "post" if post else "get")(
+                    url,
+                    # headers=self.return_headers(url),
```
Comment (joshmaerk) [inline safaribooks.py]: headers has no effect anymore.

File: safaribooks.py [hunk 4]
```diff
@@ -384,12 +386,15 @@ def update_cookies(self, jar):
...
     def requests_provider(self, url, post=False, data=None, update_cookies=True, **kwargs):
         try:
-            response = getattr(requests, "post" if post else "get")(
-                url,
-                headers=self.return_headers(url),
-                data=data,
-                **kwargs
-            )
+            if not self.session:
+                self.get_session(data=data)
```
Comment (joshmaerk) [inline safaribooks.py]: write session to self.session, if it not exists already

File: safaribooks.py [hunk 5]
```diff
@@ -406,6 +411,33 @@ def requests_provider(self, url, post=False, data=None, update_cookies=True, **k
 
         return response
 
+    def get_login_header(self):
+        return {"content-type": "application/json"}
```
Comment (joshmaerk) [inline safaribooks.py]: function to get explicit header in login form request

File: safaribooks.py [hunk 6]
```diff
@@ -421,41 +453,17 @@ def parse_cred(cred):
...
-        )
+            data=json.dumps({"email": email, "password": str(password)}),
+            )
 
         if response == 0:
             self.display.exit("Login: unable to perform auth to Safari Books Online.\n    Try again...")
 
-        if response.status_code != 302:
+
+        if response.status_code != 200:
```
Comment (joshmaerk) [inline safaribooks.py]: Status Code is new on 200

**Comment by casperbh96** [comment]
I tried this to fix _no CSRF Token found_, but now I get another error.

```
[20/May/2019 01:47:02] ** Welcome to SafariBooks! **
[20/May/2019 01:47:02] Logging into Safari Books Online...
[20/May/2019 01:47:03]   File "safaribooks.py", line 1035, in <module>
    SafariBooks(args_parsed)
  File "safaribooks.py", line 289, in __init__
    self.do_login(*args.cred)
  File "safaribooks.py", line 459, in do_login
    data=json.dumps({"email": email, "password": str(password)}),
  File "safaribooks.py", line 390, in requests_provider
    self.get_session(data=data)
  File "safaribooks.py", line 430, in get_session
    s.get(json.loads(login_response.content)['redirect_uri'])

[20/May/2019 01:47:03] Unhandled Exception: 'redirect_uri' (type: KeyError)

```

**Comment by SaschaJohn** [comment]
@joshmaerk 
for me the fix does not work

λ python --version
Python 3.5.1

[#] Unhandled Exception: the JSON object must be str, not 'bytes' (type: TypeError)

to fix, I changed line 430 and added a decode():
s.get(json.loads(login_response.content.decode())['redirect_uri'])

afterwards I get a:
[#] API: Out-of-Session (Authentication credentials were not provided.).

**Comment by mondrias** [comment]
Work in Windows with anaconda thanks!

**Comment by nathanernest** [comment]
This works. Thanks @joshmaerk

**Comment by davistran86** [comment]
@joshmaerk I got error when it starts to download book CSSs:

```
Downloading book CSSs... (4 files)
Process Process-3: 
Process Process-2:
Traceback (most recent call last): 
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs) 
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "safaribooks.py", line 790, in _thread_download_css
    response = self.requests_provider(url, update_cookies=False)
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "safaribooks.py", line 396, in requests_provider
    **kwargs
  File "safaribooks.py", line 790, in _thread_download_css 
    response = self.requests_provider(url, update_cookies=False)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "safaribooks.py", line 396, in requests_provider
    **kwargs 
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/sessions.py", line 646, in send 
    r = adapter.send(request, **kwargs)
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/davistran86/.local/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request        
    six.raise_from(e, None) 
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "<string>", line 2, in raise_from
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request        
    six.raise_from(e, None) 
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request        
    httplib_response = conn.getresponse()
  File "<string>", line 2, in raise_from
  File "/usr/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request        
    httplib_response = conn.getresponse() 
  File "/usr/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1") 
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 294, in recv_into
    return self.connection.recv_into(*args, **kwargs)
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1717, in recv_into
    self._raise_ssl_error(self._ssl, result)
  File "/home/davistran86/.local/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 294, in recv_into 
    return self.connection.recv_into(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1548, in _raise_ssl_error
    _raise_current_error()
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1717, in recv_into
    self._raise_ssl_error(self._ssl, result) 
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1548, in _raise_ssl_error
    _raise_current_error()
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'decryption failed or bad record mac')]
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors) 
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'wrong version number')]`

**Comment by ghnu** [comment]
> I tried this to fix _no CSRF Token found_, but now I get another error.
> 
> ```
> [20/May/2019 01:47:02] ** Welcome to SafariBooks! **
> [20/May/2019 01:47:02] Logging into Safari Books Online...
> [20/May/2019 01:47:03]   File "safaribooks.py", line 1035, in <module>
>     SafariBooks(args_parsed)
>   File "safaribooks.py", line 289, in __init__
>     self.do_login(*args.cred)
>   File "safaribooks.py", line 459, in do_login
>     data=json.dumps({"email": email, "password": str(password)}),
>   File "safaribooks.py", line 390, in requests_provider
>     self.get_session(data=data)
>   File "safaribooks.py", line 430, in get_session
>     s.get(json.loads(login_response.content)['redirect_uri'])
> 
> [20/May/2019 01:47:03] Unhandled Exception: 'redirect_uri' (type: KeyError)
> ```

Same here. There is now another error that needs to be fixed to get it working again.

@joshmaerk : Please help.

**Comment by lorenzodifuccia** [comment]
Please, review the code and try to adapt it with the latest commit.
See #144.

**Comment by lorenzodifuccia** [comment]
Closed with 31d7c9.
Thanks for the support.

**Pull request closed without merging**